### PR TITLE
test(quantic): add playwright tests for result quickview

### PR DESCRIPTION
**Pull request by mikegron**
## [SFINT-6016](https://coveord.atlassian.net/browse/SFINT-6016)

Quantic Result Quickview playwright tests!

[SFINT-6016]: https://coveord.atlassian.net/browse/SFINT-6016?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

image

image

### CODE REVIEW

**Review by SimonMilord — COMMENTED** [review]
Can you add a screenshot for your test run in the PR description? You can use the --ui flag in the command to open the UI for your screenshot. It is quite useful when reviewing to :

1- easily see the test structure
2- see the performance of these tests.

File: packages/quantic/playwright/utils/requests.ts [hunk 1]
```diff
@@ -13,6 +13,9 @@ export const analyticsSearchesUrlRegex =
 export const analyticsCustomUrlRegex = /\/rest\/v15\/analytics\/custom/;
 export const analyticsClickUrlRegex = /\/rest(\/ua)?\/v15\/analytics\/click/;
 export const analyticsCollectUrlRegex = /\/rest\/v15\/analytics\/collect/;
+export const searchHtmlRequestRegex = /\/rest\/search\/v2\/html.*/;
```
Comment (SimonMilord) [inline packages/quantic/playwright/utils/requests.ts:16]:
```suggestion
export const searchQuickviewRequestRegex = /\/rest\/search\/v2\/html.*/;
```

just a suggestion

File: packages/quantic/playwright/page-object/searchObjectWithQuickview.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import {Page} from '@playwright/test';
+import {SearchObject} from './searchObject';
+
+export class SearchObjectWithQuickview extends SearchObject {
+  constructor(
+    page: Page,
+    searchRequestRegex: RegExp,
+    private documentHtmlRequestRegex: RegExp
+  ) {
+    super(page, searchRequestRegex);
```
Comment (SimonMilord) [inline packages/quantic/playwright/page-object/searchObjectWithQuickview.ts:10]: here we usually use the pattern : `this.page = page;` instead of the super

File: packages/quantic/force-app/examples/main/lwc/exampleQuanticResultQuickview/exampleQuanticResultQuickview.html [hunk 3]
```diff
@@ -9,7 +9,7 @@
             <c-action-perform-search slot="actions" engine-id={engineId}></c-action-perform-search>
         </c-configurator>
 
-        <c-quantic-search-interface slot="preview" engine-id={engineId}>
+        <c-example-use-case slot="preview" use-case={config.useCase} engine-id={engineId}>
```
Comment (SimonMilord) [inline packages/quantic/force-app/examples/main/lwc/exampleQuanticResultQuickview/exampleQuanticResultQuickview.html:12]: 👍

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/fixture.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        new SearchObjectWithQuickview(
+          page,
+          searchRequestRegex,
+          searchHtmlRequestRegex
+        )
+      );
+    },
+    resultQuickview: async ({page, options, configuration, urlHash}, use) => {
+      await page.goto(
+        urlHash ? `${resultQuickviewUrl}#${urlHash}` : resultQuickviewUrl
```
Comment (SimonMilord) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/fixture.ts:55]: I dont think you need urlHash here as I dont see any test that require checking or passing something in the url

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/pageObject.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    return this.page.locator('c-quantic-result-quickview button');
+  }
+
+  get buttonIcon(): Locator {
+    return this.button.locator('lightning-icon');
+  }
+
+  get buttonTooltip(): Locator {
+    return this.button
+      .locator('..')
```
Comment (SimonMilord) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/pageObject.ts:18]: ? why the .locator('..') ?

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/pageObject.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    this.page = page;
+  }
+
+  get button(): Locator {
+    return this.page.locator('c-quantic-result-quickview button');
+  }
+
+  get buttonIcon(): Locator {
+    return this.button.locator('lightning-icon');
+  }
```
Comment (SimonMilord) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/pageObject.ts:14]:
```suggestion
  get quickviewButton(): Locator {
    return this.page.locator('c-quantic-result-quickview button');
  }

  get quickviewButtonIcon(): Locator {
    return this.quickviewButton.locator('lightning-icon');
  }
```

Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/pageObject.ts:14]: I would rename them quickviewButton, etc. to make it more precise and clear what we are selecting here.

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/pageObject.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+  get buttonTooltip(): Locator {
+    return this.button
+      .locator('..')
+      .locator('c-quantic-tooltip div[slot="content"]');
+  }
+
+  get modalTitle(): Locator {
```
Comment (SimonMilord) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/pageObject.ts:22]: same here maybe something like "quickviewModalTitle" and so on

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/quanticResultQuickView.e2e.ts
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/quanticResultQuickView.e2e.ts:1]: One important thing to keep in mind when creating these tests is the following document: https://github.com/coveo/ui-kit/blob/master/packages/quantic/decisions/0001-testing-strategy.md
It's our strategy regarding what to test in Quantic with E2E tests and what to test in unit tests.

For example in these tests I see some tests that should be unit tests instead of E2E tests, additionally we need to test sending analytics with these E2E tests.
To summarize the strategy:

- Jest Unit Tests should focus on isolated component behaviour, mocking external dependencies, validating HTML rendering, property updates, and testing edge cases. These tests should not involve full user workflows or real external systems.

- Playwright E2E Tests should simulate user interactions and ensure integration with external systems like APIs, Salesforce, and analytics. These tests verify that components work as expected within the full UI environment.

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/quanticResultQuickView.e2e.ts [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        await expect(contentContainer).toBeVisible();
+        const iframe = contentContainer.locator('iframe');
+        const iframeContent = await iframe.contentFrame();
+        await expect(iframeContent?.locator('div')).toContainText(
+          'HTML content'
+        );
+      });
+    });
+
+    test.describe('when result has no html version', () => {
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/quanticResultQuickView.e2e.ts:44]: I believe this could be a unit test.

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/quanticResultQuickView.e2e.ts [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+      });
+
+      test('should disable quickview button', async ({resultQuickview}) => {
+        const previewButton = resultQuickview.button;
+        await expect(previewButton).toBeVisible();
+        await expect(previewButton).toBeDisabled();
+      });
+    });
+
+    test.describe('with custom preview button icon', () => {
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/quanticResultQuickView.e2e.ts:58]: I believe this could be a unit test.

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/quanticResultQuickView.e2e.ts [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        const previewButton = resultQuickview.button;
+        await expect(previewButton).toBeVisible();
+
+        const buttonIcon = resultQuickview.buttonIcon;
+        await expect(buttonIcon).toBeVisible();
+        await expect(buttonIcon).toHaveAttribute('icon-name', customIcon);
+      });
+    });
+
+    test.describe('with custom label', () => {
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/quanticResultQuickView.e2e.ts:76]: I believe this could be a unit tests.

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/quanticResultQuickView.e2e.ts [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+        },
+      });
+
+      test('should display the label', async ({resultQuickview}) => {
+        const label = resultQuickview.button;
+        await expect(label).toContainText(customLabel);
+      });
+    });
+
+    test.describe('with custom tooltip', () => {
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/quanticResultQuickView.e2e.ts:90]: I believe this could be a unit test.

**Review by mmitiche — COMMENTED** [review]
We should've mentioned this strategy earlier: https://github.com/coveo/ui-kit/blob/master/packages/quantic/decisions/0001-testing-strategy.md
Sorry about that.

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/pageObject.ts (hunk 5 above)
Comment (mikegron) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/pageObject.ts:18]: J'avais besoin d'aller dans le parent, mais je pourrais faire un autre getter sinon!

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/pageObject.ts (hunk 6 above)
Comment (mikegron) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/pageObject.ts:14]: Je suis pas tant d'accord vue que l'objet est déja un `ResultQuickviewObject`, dans les tests `resultQuickview.button` look mieux que `resultQuickview.quickviewButton` ?

Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/pageObject.ts:14]: c'est un matter de preference pas un deal breaker pour moi, dans les autres test files c'est le pattern qu'on follow

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/__tests__/quanticResultQuickview.test.js [hunk 12]
```diff
@@ -131,41 +135,158 @@ describe('c-quantic-result-quick-view', () => {
...
+        selectors.quickviewButton
+      );
+
+      expect(quickViewButton.textContent).toContain('Custom Label');
+      expect(quickViewButton.querySelector(selectors.icon).classList).toContain(
+        'slds-button__icon_right'
+      );
+    });
+
+    it.each([['brand'], ['outline_brand'], ['result-action'], ['turbo']])(
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/__tests__/quanticResultQuickview.test.js:206]: what is this `turbo` variant?

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/__tests__/quanticResultQuickview.test.js
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/__tests__/quanticResultQuickview.test.js:1]: I see there are couple more other things that we can cover:
- the logic of dispatching a `quantic__resultactionregister` event when the variant used is `'result-action'`  : https://github.com/coveo/ui-kit/blob/cc7d962cd07f23429091d86cf974c15ef4e98209/packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.js#L111-L123

- correct initialization of the component and subscribing to the state, example in another component: https://github.com/coveo/ui-kit/blob/cc7d962cd07f23429091d86cf974c15ef4e98209/packages/quantic/force-app/main/default/lwc/quanticGeneratedAnswer/__tests__/quanticGeneratedAnswer.test.js#L239-L279

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/__tests__/quanticResultQuickview.test.js [hunk 13]
```diff
@@ -131,41 +135,158 @@ describe('c-quantic-result-quick-view', () => {
...
+
+      expect(element.maximumPreviewSize).toBe(500);
     });
   });
 
-  describe('when the component is not used in the search use case', () => {
-    beforeAll(() => {
-      useCase = 'insight';
+  describe('when the quickview is opened and closed', () => {
+    it('should open the quickview modal', async () => {
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/__tests__/quanticResultQuickview.test.js:254]: When of the things that could be useful to test is to confirm that the `engineId` property used when creating the QuanticResultQuickview component has been successfully passed to the QuanticResultLink component displayed when opening the Quickview.

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/quanticResultQuickView.e2e.ts
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/quanticResultQuickView.e2e.ts:1]: One of the additional things that needs to be tested in E2E is the display of the `footer` slot in the Quick view modal: https://github.com/coveo/ui-kit/blob/cc7d962cd07f23429091d86cf974c15ef4e98209/packages/quantic/force-app/main/default/lwc/quanticResultQuickview/quanticResultQuickview.html#L51

we need to make sure that when this slot is passed, it is correctly rendered.
We recently got a bug related to that: https://github.com/coveo/ui-kit/pull/5090

**Review by erocheleau — COMMENTED** [review]
Very nice, slight suggestions, also Mehdi has a lot of very good comments! :+1:

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/__tests__/quanticResultQuickview.test.js [hunk 14]
```diff
@@ -131,41 +135,158 @@ describe('c-quantic-result-quick-view', () => {
...
+
+    it('should render the maximum preview size correctly', async () => {
+      const customOptions = {
+        result: exampleResult,
+        maximumPreviewSize: 500,
+      };
+      const element = createTestComponent(customOptions);
+      await flushPromises();
+
+      expect(element.maximumPreviewSize).toBe(500);
```
[the remark this answers is not in the document]
Comment (mikegron) [reply packages/quantic/force-app/main/default/lwc/quanticResultQuickview/__tests__/quanticResultQuickview.test.js:249]: ta ben raison lol j'ai skip ca trop vite

**Review by mmitiche — APPROVED** [review]
LGTM! nice job!

File: packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/pageObject.ts [hunk 15]
```diff
@@ -0,0 +1,38 @@
+import type {Locator, Page, Request} from '@playwright/test';
+import {isUaClickEvent} from '../../../../../../playwright/utils/requests';
+
+export class ResultQuickviewObject {
+  constructor(public page: Page) {
+    this.page = page;
+  }
+
+  get quickviewButton(): Locator {
+    return this.page.locator('c-quantic-result-quickview button');
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticResultQuickview/e2e/pageObject.ts:10]: If possible please try with a `getByRole` as much as possible for buttons selectors.
This also helps us with accessibility concerns because the `getByRole` also represents some of the pre-requisites for the buttons to be picked up correctly by screen readers for example.

(But sometimes we can't get it to work in Salesforce, so if it doesn't work, just leave as-is and reply to this comment)

**Review by erocheleau — APPROVED** [review]
Just 1 comment to try and use getByRole as a locator for buttons.
But I'll approve anyway, if you can make it work, please change it, if you can't it's okay.
